commit 4fa81ed27781a12f6303b9263056635ae74e3e21
Author: Martin Schwidefsky <schwidefsky@de.ibm.com>
Date:   Wed Mar 18 13:27:32 2009 +0100

    [S390] __div64_31 broken for CONFIG_MARCH_G5
    
    The implementation of __div64_31 for G5 machines is broken. The comments
    in __div64_31 are correct, only the code does not do what the comments
    say. The part "If the remainder has overflown subtract base and increase
    the quotient" is only partially realized, the base is subtracted correctly
    but the quotient is only increased if the dividend had the last bit set.
    Using the correct instruction fixes the problem.
    
    Cc: stable@kernel.org
    Reported-by: Frans Pop <elendil@planet.nl>
    Tested-by: Frans Pop <elendil@planet.nl>
    Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

diff --git a/arch/s390/lib/div64.c b/arch/s390/lib/div64.c
index a5f8300..d9e62c0 100644
--- a/arch/s390/lib/div64.c
+++ b/arch/s390/lib/div64.c
@@ -61,7 +61,7 @@ static uint32_t __div64_31(uint64_t *n, uint32_t base)
 		"	clr	%0,%3\n"
 		"	jl	0f\n"
 		"	slr	%0,%3\n"
-		"	alr	%1,%2\n"
+		"	ahi	%1,1\n"
 		"0:\n"
 		: "+d" (reg2), "+d" (reg3), "=d" (tmp)
 		: "d" (base), "2" (1UL) : "cc" );
